src/kbmod/analysis_utils.py

```diff
@@ -931,105 +758,82 @@ def apply_stamp_filter(
                     if stamp.get_width() > 1:
-                        keep["stamps"].append(np.array(stamp))
+                        result_list.results[ind + start_idx].stamp = np.array(stamp)
                         all_valid_inds.append(ind + start_idx)
 
                 # Move to the next chunk.
                 start_idx += chunk_size
+                del stamps_slice
 
-        if len(keep2) > 0:
-            passing_stamps_idx = np.unique(np.concatenate(passing_stamps_idx))
-            keep2 = np.concatenate(keep2, axis=0)
-
-        keep["final_results"] = np.array(all_valid_inds)
-        keep["stamps"] = np.array(keep["stamps"])
-        print("Keeping %i results" % len(keep["final_results"]), flush=True)
+        # Do the actual filtering of results
+        result_list.filter_results(all_valid_inds)
+        print("Keeping %i results" % result_list.num_results(), flush=True)
 
         end_time = time.time()
         time_elapsed = end_time - start_time
         print("{:.2f}s elapsed".format(time_elapsed))
-        return keep
 
-    def apply_clustering(self, keep, cluster_params):
+    def apply_clustering(self, result_list, cluster_params):
         """
         This function clusters results that have similar trajectories.
 
         Arguments:
-            keep (dict): Contains the values of which results were kept from the
-                search algorithm
-            cluster_params (dict): Contains values concerning the image and
-                search initial settings including: x_size, y_size, vel_lims,
-                ang_lims, and mjd.
-
-        Returns:
-            dict: A results dictionary as generated by SharedTools.gen_results_dict.
+            result_list (`ResultList`): The values from trajectories. This data gets
+                modified directly by the filtering.
+            cluster_params (dict): Contains values concerning the image and search
+                settings including: x_size, y_size, vel_lims, ang_lims, and mjd.
         """
-        rs = ResultList(cluster_params["mjd"])
-        rs.append_result_dict(keep)
-
         # Skip clustering if there is nothing to cluster.
-        if rs.num_results() == 0:
-            return keep
-        print("Clustering %i results" % rs.num_results(), flush=True)
+        if result_list.num_results() == 0:
+            return
+        print("Clustering %i results" % result_list.num_results(), flush=True)
 
         # Do the clustering and the filtering.
         cluster_idx = self._cluster_results(
-            np.array(rs.trajectory_list()),
+            np.array(result_list.trajectory_list()),
             cluster_params["x_size"],
             cluster_params["y_size"],
             cluster_params["vel_lims"],
             cluster_params["ang_lims"],
             cluster_params["mjd"],
         )
-        rs.filter_results(cluster_idx)
+        result_list.filter_results(cluster_idx)
         del cluster_idx
```

Comment:
I'm not sure we need this either, or is there something about Python's GC that makes this so memory intensive this is required?

Reply:
Removed.